### PR TITLE
Support Energon as dataloader in NeVA

### PR DESCRIPTION
**Pull request by paul-gibbons**
# What does this PR do ?
This PR adds experimental support for utilizing [Megatron-Energon](https://github.com/NVIDIA/Megatron-Energon) as a dataloader for NeVA training.

Co-authored:
@xuanzic 

**Collection**: [Note which collection this PR will affect]

# Changelog 
- Add specific line by line info of high level changes in this PR.

# Usage
* You can potentially add a usage example below

```python
# Add a code snippet demonstrating how to use this 
```

# GitHub Actions CI

The Jenkins CI system has been replaced by GitHub Actions self-hosted runners.

The GitHub Actions CI will run automatically when the "Run CICD" label is added to the PR.
To re-run CI remove and add the label again.
To run CI on an untrusted fork, a NeMo user with write access must first click "Approve and run".

# Before your PR is "Ready for review"
**Pre checks**:
- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [ ] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

File: examples/multimodal/multimodal_llm/neva/conf/neva_config.yaml [hunk 1]
```diff
@@ -210,6 +210,23 @@ model:
     image_folder: null
     image_aspect_ratio: 'square'
 
+energon:
+  use_energon: False
+  data:
+    __module__: megatron.energon
+    __class__: Metadataset
```
Comment (yaoyu-33) [inline examples/multimodal/multimodal_llm/neva/conf/neva_config.yaml:217]: can you add a bit more comments here? explain the usage of __class__ and the weights below?

File: nemo/collections/multimodal/data/neva/neva_energon_dataset.py [hunk 2]
```diff
@@ -0,0 +1,561 @@
+import dataclasses
```
[the remark this answers is not in the document]
Comment (yaoyu-33) [reply nemo/collections/multimodal/data/neva/neva_energon_dataset.py:1]: please fix code scan comments

File: nemo/collections/multimodal/models/multimodal_llm/neva/neva_model.py [hunk 3]
```diff
@@ -1435,6 +1463,143 @@ def build_pretraining_data_loader(
             persistent_workers=True if self.cfg.data.num_workers > 0 else False,
         )
 
+    def datasets_provider(self, worker_config=None):
```
Comment (yaoyu-33) [inline nemo/collections/multimodal/models/multimodal_llm/neva/neva_model.py:1466]: is it possible to move this method outside the model class?

File: nemo/collections/multimodal/data/neva/neva_energon_dataset.py [hunk 4]
```diff
@@ -0,0 +1,480 @@
...
+    preprocess_nv_dpo,
+    preprocess_nvgpt,
+    preprocess_plain,
+    preprocess_v1,
+    process_image,
+)
+
+
+@dataclass
+class ImageTaskSample:
```
Comment (yaoyu-33) [inline nemo/collections/multimodal/data/neva/neva_energon_dataset.py:40]: add some comments, if this is required from mcore

File: nemo/collections/multimodal/data/neva/neva_energon_dataset.py [hunk 5]
```diff
@@ -0,0 +1,480 @@
...
+
+        # TODO: cleanup, this is following logic in neva_dataset when we rearrange media tensor
+        if batch.media.shape[1] == 1:
+            batch.media = rearrange(batch.media, "b T c h w -> b T 1 c h w")
+        else:
+            batch.media = rearrange(batch.media, "b T c h w -> b T 1 c h w")
+
+        return batch
+
+    def preprocess_conversations(self, sources):
```
Comment (yaoyu-33) [inline nemo/collections/multimodal/data/neva/neva_energon_dataset.py:426]: any reason this part cannot be shared with neva_dataset?

File: nemo/collections/multimodal/data/neva/neva_energon_dataset.py [hunk 6]
```diff
@@ -0,0 +1,480 @@
...
+class ImageTaskBatch(Batch):
+    tokens: torch.Tensor
+    labels: torch.Tensor
+    attention_mask: torch.Tensor
+    loss_mask: torch.Tensor
+    position_ids: torch.Tensor
+    media: Optional[torch.Tensor] = None
+
+
+class TaskEncoder(DefaultTaskEncoder[VQASample, InterleavedSample, ImageTaskBatch, dict]):
```
Comment (yaoyu-33) [inline nemo/collections/multimodal/data/neva/neva_energon_dataset.py:64]: add docstring plz

**Review by yaoyu-33 — DISMISSED** [review]

**Review by yaoyu-33 — APPROVED** [review]
